File: 프로그래머스/키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    leftSpot = "*"
    rightSpot = "#"
    # 키패드 숫자를 좌표로 표현
    position = {1 : (0,0), 2 : (0,1), 3 : (0,2), 4 : (1,0), 5 : (1,1), 6 : (1,2),
               7 : (2,0), 8 : (2,1), 9 : (2,2), "*" : (3,0), 0 : (3,1), "#" : (3,2)}
        
    for n in numbers:
        if n == 1 or n == 4 or n == 7:
            answer += "L"
            leftSpot = n
        elif n == 3 or n == 6 or n == 9:
            answer += "R"
            rightSpot = n
        else:
            # 가로 좌표, 세로 좌표 차이 각각 구해서 더하면 총거리 차이
            l = abs(position[leftSpot][0] - position[n][0]) + abs(position[leftSpot][1] - position[n][1])
            r = abs(position[rightSpot][0] - position[n][0]) + abs(position[rightSpot][1] - position[n][1])
            
            if l < r:
                answer += "L"
                leftSpot = n
            elif l > r:
                answer += "R"
                rightSpot = n
            else:
                if hand == "left":
                    answer += "L"
                    leftSpot = n
                else:
                    answer += "R"
                    rightSpot = n
            
    return answer

File: 프로그래머스/키패드 누르기.java
public String solution(int[] numbers, String hand) {
        StringBuilder sb = new StringBuilder();
        // 오른손잡이 = true
        boolean prefRight = hand.equals("right");
        // 왼손, 오른손 엄지손가락의 좌표 위치를 나타내는 변수(처음엔 *,# 위치)
        int[] leftPos = {4, 1}, rightPos = {4, 3};
        // 모든 숫자(0~9)의 좌표 위치를 HashMap에 등록
        int[][] numLoc = {{4, 2}, {1, 1}, {1, 2}, {1, 3}, {2, 1}, {2, 2}, {2, 3}, {3, 1}, {3, 2}, {3, 3}};

        for (int num : numbers) {
            switch (num) {
                case 1: case 4: case 7:
                    sb.append('L');
                    leftPos = numLoc[num];
                    break;
                case 3: case 6: case 9:
                    sb.append('R');
                    rightPos = numLoc[num];
                    break;
                default:
                    int leftDis = Math.abs(numLoc[num][0] - leftPos[0]) + Math.abs(numLoc[num][1] - leftPos[1]);
                    int rightDis = Math.abs(numLoc[num][0] - rightPos[0]) + Math.abs(numLoc[num][1] - rightPos[1]);
                    if (leftDis > rightDis) {
                        sb.append('R');
                        rightPos = numLoc[num];
                    } else if (rightDis > leftDis) {
                        sb.append('L');
                        leftPos = numLoc[num];
                    } else {
                        if (prefRight) {
                            sb.append('R');
                            rightPos = numLoc[num];
                        } else {
                            sb.append('L');
                            leftPos = numLoc[num];
                        }
                    }
            }
        }
        return sb.toString();
    }